### PR TITLE
Add events integration to Incremental Entity Providers

### PR DESCRIPTION
**Pull request by dekoding**
## Hey, I just made a Pull Request!

This pull request adds a new feature to incremental entity providers: The ability to receive events and (optionally) trigger a delta update based on an event payload. A new, optional `eventHandler` property exposes methods for handling events and identifying event topics that should be handled by the provider:

```typescript
export interface IncrementalEntityProvider<TCursor, TContext> {
  ...
  /**
   * If set, the IncrementalEntityProvider will receive and respond to
   * events.
   *
   * This system acts as a wrapper for the Backstage events bus, and
   * requires the events backend to function. It does not provide its
   * own events backend. See {@link https://github.com/backstage/backstage/tree/master/plugins/events-backend}.
   */
  eventHandler?: {
    /**
     * This method accepts an incoming event for the provider, and
     * optionally maps the payload to an object containing a delta
     * mutation.
     *
     * If a valid delta is returned by this method, it will be ingested
     * automatically by the provider.
     */
    onEvent: (params: EventParams) =>
      | undefined
      | {
          added: DeferredEntity[];
          removed: { entityRef: string }[];
        };

    /**
     * This method returns an array of topics for the IncrementalEntityProvider
     * to respond to.
     */
    supportsEventTopics: () => string[];
  };
}
```

This feature is optional and should not break any existing providers.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by freben** [comment]
I'm a bit confused by this one. It seems like this makes the engine do two completely distinct things - pull data from external sources and pushing them inward, and also receiving events and purely translating them to additions/removals via a callback. What drove this change, and are the two more related than they seem at surface value?

If the added functionality is needed, I'd have thought it might be in a provider all of its own, and perhaps in a more targeted or system specific way, rather than a generic translation by callback to the raw entity provider interface 🤔

**Comment by dekoding** [comment]
> I'm a bit confused by this one. It seems like this makes the engine do two completely distinct things - pull data from external sources and pushing them inward, and also receiving events and purely translating them to additions/removals via a callback. What drove this change, and are the two more related than they seem at surface value?
> 
> If the added functionality is needed, I'd have thought it might be in a provider all of its own, and perhaps in a more targeted or system specific way, rather than a generic translation by callback to the raw entity provider interface thinking

Good question. The logic behind giving incremental providers the ability to perform on-the-fly delta updates like this is to allow for situations where new assets are added to a data source that is capable of performing notifications (i.e., webhooks) outside the ingestion window.

For example, let's say you're using an incremental provider to ingest data for fifty thousand repositories from Github Enterprise. Since most won't change over the course of a week, you only have it scheduled to perform incremental ingestion once a week. However, you are aware that a repo might be added or removed during the rest period for the incremental ingestion, and GHE offers webhooks you can subscribe to in order to get those atomic updates. If the provider also supports deltas, you can get those on-the-fly updates ingested practically instantly. And the next time regular incremental ingestion runs, it will pick up those changes and incorporate them into the `ingestion_mark_entities` table.

As for reasons why not to use a separate entity provider, the main one is that if the entity is created by a separate `EntityProvider`, it will be managed by that provider, rather than the one at the core of the incremental provider. If the incremental provider subsequently attempts to ingest the same entity with the same entityref, you'll get collisions. Unless there's a way to "hand off" an entity from one provider to another that I'm not aware of?

File: plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts [hunk 1]
```diff
@@ -326,4 +331,42 @@ export class IncrementalIngestionEngine implements IterationEngine {
...
+      return;
+    }
+
+    const update = provider.deltaMapper(payload);
+
+    if (update.delta) {
+      await connection.applyMutation({
+        type: 'delta',
+        ...update.delta,
+      });
```
Comment (freben) [inline plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts:358]: I want to emphasize that it's important that whatever this provider does send over the connection, it needs to record in its own tables as well. If a webhook leads to an add here, and then during incremental ingestion it's no longer there, it won't be removed again unless we're recorded this fact when we pushed here.

**Comment by freben** [comment]
Regarding overwrites: providers can actually overwrite each other, as long as they do so using the exact same `locationKey`. Of course, it's important that they don't do so with completely different entity shapes each time because that would be very confusing and lead to churn :) But let's see if we keep these as one anyway. 👇 

Going to ping @pjungermann here for the interesting event use case.

Regarding the pull + push duality:

Alright, thanks for the context! This is indeed something that will be important, but we've thought about it from the opposite angle so far. We'd have expected the world to gradually shift towards being _primarily_ event based and centred around that, and then pull mechanisms will still linger around but at very slow rates and only basically for fault recovery (making sure that stuff that for some reason falls between the chairs eventually "heals" anyway). That makes them "secondary" in a sense.

This makes me think that maybe a provider that has this ability, and being fed primarily out of webhooks, might want to be an event handler first, and then maybe, but as a side concern, it might be using the incremental method of ingestion in the recovery mode. :) This makes me look at your original stab with incremental ingestion more as a "facility" than a provider in a different light.

Maybe it's a matter of just tweaking the interfaces here a little, to clarify this. The name `deltaMapper` is the main offender perhaps. It would be good if one could read the `IncrementalEntityProvider` interface from top to bottom and it were clear that there was an entity bus part that subscribes to a special topic, and then there was a scheduled static part.

The interface could have optional `supportsEventTopic` and `onEvent` methods or something. Maybe if there's an onEvent but no supportsEventTopic, it'll fall back to a topic with the same name as the provider? 🤷 maybe that's more confusing than helpful.

And then be prepared that maybe people will more commonly come from the angle of having written event based ingestion first elsewhere, and only later, when they feel that they need to perhaps have an incremental fallback, do they look at this system and try to rewrite their provider as an incremental one.

File: plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts (hunk 1 above)
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts:358]: Sorry for the delay, things got busy. I'm *NOW* going to work on this. Also, it occurred to me that I was making deletions too complicated. The end goal, no matter which logical path, is to remove the reference to the entity from the `ingestion_mark_entities` table if we've received a delta for `applyMutation` to run, so all that needs to happen is removal.

Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts:387]: And done!

**Comment by pjungermann** [comment]
_I'm already on my Christmas / end of year vacation and hence, less active right now._

**Disclaimer:** I know more or less what the incremental entity provider is, however I have no knowledge about how it works in detail.

And I didn't look at the code changes yet. I will have a closer look later. Just did a super quick scroll through it.

> We'd have expected the world to gradually shift towards being primarily event based and centered around that, and then pull mechanisms will still linger around but at very slow rates and only basically for fault recovery (making sure that stuff that for some reason falls between the chairs eventually "heals" anyway). That makes them "secondary" in a sense.

I fully agree to this statement.

> We want to get there, too. We regard incremental ingestion as kind of a drop-in replacement for full mutations for those cases where full is impossible due to the size of the entity set [...]

Overall, I would see the incremental ingestion as a replacement for the full mutation, too.

"drop-in replacement" means for me that there is nearly no additional work, however I assume that it is not that simple. A level of encapsulation and abstraction which allows you to use the capability without detailed knowledge is likely what you have prepared already though.

In general, I would see this as a feature for full mutations which entity providers can utilize -- or not. Maybe it even gets the default at some point with small datasets being ingested as one batch.

---
_scenario:_

E.g., let's assume there is a `FooEntityProvider` to ingest entities based on the external system `Foo`.

`Foo` provides APIs which we can use to scan and discover information which we can convert to entities.

Additionally, `Foo` provides events for changes to this data and we can subscribe to these events to refresh affected entities. In order to do so, we need to be able to identify these "affected entities" from the information within the event as well as the information we have at the provider-owned entities (name, annotations, ...) -- including expected new state and previous state (add/update/remove).

Full mutations will be scheduled in very long intervals to recover from any potentially lost event, etc.

The identification of "affected entities" (old vs new state) requires a certain level of alignment with the full mutation and its outcome. ("A" resulted in entities "a", "b", "c"; event for changes at "A"; "A" will result in the new state "a", "c", "d"; outcome: remove "a", add "d").
_(maybe "refresh keys", annotations, ...)_

As the dataset and the resulting ingested set of entities is huge (e.g., filters for the API will not be sufficient), we need a more efficient full mutation/ingestion compared to scanning the whole dataset at once. Hence, we add incremental ingestion support to the entity provider.

The incremental ingestion facilitates going through the entire dataset in batches/increments. In order to do so, it needs to keep track of the dataset.

This might require being aware of the delta mutations issued as result of event-based updates.

_You can replace `Foo` with `Github`, `BitbucketCloud`, etc. and consider their (repo) push events._

---

```ts
this.providerEventTopic = `${options.provider.getProviderName()}-delta`;
```

Not sure if events like this make sense though. I would rather expect to have something like `github.push` or `bitbucketCloud.repo:push` as consumed events.

Maybe the incremental ingestion engine just needs a way to hook into the delta mutations itself. (E.g. by wrapping the connection at `connect(connection: EntityProviderConnection): Promise<void>` and add pre and/or post actions to `applyMutation` in case of mutation.type delta).

As written above: take these preliminary thoughts with a grain of salt. :-)

**Comment by dekoding** [comment]
> _I'm already on my Christmas / end of year vacation and hence, less active right now._
> 
> **Disclaimer:** I know more or less what the incremental entity provider is, however I have no knowledge about how it works in detail.
> 
> And I didn't look at the code changes yet. I will have a closer look later. Just did a super quick scroll through it.
> 
> > We'd have expected the world to gradually shift towards being primarily event based and centered around that, and then pull mechanisms will still linger around but at very slow rates and only basically for fault recovery (making sure that stuff that for some reason falls between the chairs eventually "heals" anyway). That makes them "secondary" in a sense.
> 
> I fully agree to this statement.
> 
> > We want to get there, too. We regard incremental ingestion as kind of a drop-in replacement for full mutations for those cases where full is impossible due to the size of the entity set [...]
> 
> Overall, I would see the incremental ingestion as a replacement for the full mutation, too.
> 
> "drop-in replacement" means for me that there is nearly no additional work, however I assume that it is not that simple. A level of encapsulation and abstraction which allows you to use the capability without detailed knowledge is likely what you have prepared already though.
> 
> In general, I would see this as a feature for full mutations which entity providers can utilize -- or not. Maybe it even gets the default at some point with small datasets being ingested as one batch.
> 
> _scenario:_
> 
> E.g., let's assume there is a `FooEntityProvider` to ingest entities based on the external system `Foo`.
> 
> `Foo` provides APIs which we can use to scan and discover information which we can convert to entities.
> 
> Additionally, `Foo` provides events for changes to this data and we can subscribe to these events to refresh affected entities. In order to do so, we need to be able to identify these "affected entities" from the information within the event as well as the information we have at the provider-owned entities (name, annotations, ...) -- including expected new state and previous state (add/update/remove).
> 
> Full mutations will be scheduled in very long intervals to recover from any potentially lost event, etc.
> 
> The identification of "affected entities" (old vs new state) requires a certain level of alignment with the full mutation and its outcome. ("A" resulted in entities "a", "b", "c"; event for changes at "A"; "A" will result in the new state "a", "c", "d"; outcome: remove "a", add "d"). _(maybe "refresh keys", annotations, ...)_
> 
> As the dataset and the resulting ingested set of entities is huge (e.g., filters for the API will not be sufficient), we need a more efficient full mutation/ingestion compared to scanning the whole dataset at once. Hence, we add incremental ingestion support to the entity provider.
> 
> The incremental ingestion facilitates going through the entire dataset in batches/increments. In order to do so, it needs to keep track of the dataset.
> 
> This might require being aware of the delta mutations issued as result of event-based updates.
> 
> _You can replace `Foo` with `Github`, `BitbucketCloud`, etc. and consider their (repo) push events._
> 
> ```ts
> this.providerEventTopic = `${options.provider.getProviderName()}-delta`;
> ```
> 
> Not sure if events like this make sense though. I would rather expect to have something like `github.push` or `bitbucketCloud.repo:push` as consumed events.
> 
> Maybe the incremental ingestion engine just needs a way to hook into the delta mutations itself. (E.g. by wrapping the connection at `connect(connection: EntityProviderConnection): Promise<void>` and add pre and/or post actions to `applyMutation` in case of mutation.type delta).
> 
> As written above: take these preliminary thoughts with a grain of salt. :-)

Hey, no worries. I think if you look at the existing incremental provider, you'll find a lot of your thoughts are already addressed. The purpose of this particular PR is just to expose a mechanism for doing deltas with the same provider, which was previously not possible.

The mechanism I created for it takes advantage of event emitters, and it's that particular functionality I was hoping you could review if you have time.

File: plugins/catalog-backend-module-incremental-ingestion/src/types.ts [hunk 2]
```diff
@@ -154,11 +167,11 @@ export interface IterationEngine {
   taskFn: TaskFunction;
 }
 
-export interface IterationEngineOptions {
+export interface IterationEngineOptions<TInput> {
```
Comment (freben) [inline plugins/catalog-backend-module-incremental-ingestion/src/types.ts:170]: Same here, should just leave out the type argument, and having `unknown` below

File: plugins/catalog-backend-module-incremental-ingestion/src/types.ts [hunk 3]
```diff
@@ -75,6 +75,19 @@ export interface IncrementalEntityProvider<TCursor, TContext> {
    * @param burst - a function which performs a series of iterations
    */
   around(burst: (context: TContext) => Promise<void>): Promise<void>;
+
+  /**
+   * If present, this method maps incoming payloads to apply updates
+   * outside of the incremental ingestion schedule.
+   */
+  deltaMapper?: (payload: TInput) => {
```
Comment (freben) [inline plugins/catalog-backend-module-incremental-ingestion/src/types.ts:83]: I think this might be a bit too limiting.

As mentioned above, I think it'll be common that you want to listen to multiple topics (with different payloads). And the subscribers [actually receive](https://github.com/backstage/backstage/blob/d16af4a53aaa52210f8e444bb314a377aa0d868d/plugins/events-node/api-report.md?plain=1#L63) an event that has both [payload and headers](https://github.com/backstage/backstage/blob/d16af4a53aaa52210f8e444bb314a377aa0d868d/plugins/events-node/api-report.md?plain=1#L17) as well as a type, where all of the component parts probably are important in a lot of cases.

I think whatever interface we place here needs to receive raw `EventParams`, not an unpacked or simplified version of it. And it might be useful to keep the plain `supportsEventTopics(): string[]` method as-is in this interface too, and passing that through.

The `deltaMapper` method could just as well be named `onEvent` too at that point, but of course it needs to be able to return or call out to something that's specific to the incremental model.

We could return just a plain single delta as you suggest here, but I wonder if that's enough. Imagine that you listen for repo deletion events. At that point you probably need to issue some form of large sweeping deletion, and I'm not sure how to express something like that.

One possibility:

```ts
onEvent(params: EventParams): Promise<IncrementalChange> {
  return { added: ... };
}
```

Another:

```ts
onEvent(params: EventParams, ctx: SomeEventContextThing): Promise<void> {
  await ctx.delta({ added: ... });
}
```

In the latter we could add more types of "reaction" based on the event in future iterations. Not sure if premature optimization, but I was thinking that return types might be harder to evolve.

Thoughts?

File: plugins/catalog-backend-module-incremental-ingestion/api-report.md [hunk 4]
```diff
@@ -33,8 +33,8 @@ export type EntityIteratorResult<T> =
 // @public (undocumented)
 export class IncrementalCatalogBuilder {
   // (undocumented)
-  addIncrementalEntityProvider<TCursor, TContext>(
-    provider: IncrementalEntityProvider<TCursor, TContext>,
+  addIncrementalEntityProvider<TCursor, TContext, TInput>(
+    provider: IncrementalEntityProvider<TCursor, TContext, TInput>,
```
[the remark this answers is not in the document]
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/api-report.md:37]: I'll see how complicated it will be to switch to `unknown` for all of them. That would certainly make some things simpler.

I do really like the idea of making the entity provider able to handle creation/deletion/renaming. Is it appropriate for the entity _provider_ to do that, though? It seems like it might be a little out of scope.

File: plugins/catalog-backend-module-incremental-ingestion/src/types.ts (hunk 3 above)
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/types.ts:83]: I like this. Again, my only real concern is that it might be out of scope for an entity provider. But maybe not. It's almost at that point turning into an entity _manager_. But that's not really a bad thing.

File: plugins/catalog-backend-module-incremental-ingestion/api-report.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (freben) [reply plugins/catalog-backend-module-incremental-ingestion/api-report.md:37]: I'm thinking a github discovery provider that you just point at an entire org. Whenever people make new repos in there, those should also be part of the discovery. And when repos are deleted, the corresponding entities should vanish.

File: plugins/catalog-backend-module-incremental-ingestion/src/types.ts (hunk 3 above)
Comment (taras) [reply plugins/catalog-backend-module-incremental-ingestion/src/types.ts:83]: This API makes a lot of sense.

**Comment by taras** [comment]
@dekoding what do you think about renaming the issue to "Add events integration to Incremental Entity Providers"?

File: plugins/catalog-backend-module-incremental-ingestion/api-report.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/api-report.md:37]: That would be handled by the delta mechanism. Deltas can add, remove, and change entities. I thought you were suggesting that the provider actually do the work of add/remove/change, instead of processing deltas from the data source.

**Comment by dekoding** [comment]
> @dekoding what do you think about renaming the issue to "Add events integration to Incremental Entity Providers"?

That is an excellent idea. Done!

File: plugins/catalog-backend-module-incremental-ingestion/src/types.ts (hunk 3 above)
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/types.ts]: `onEvent` is actually already there! It's just in the engine, rather than in the entity provider.

Comment (freben) [reply plugins/catalog-backend-module-incremental-ingestion/src/types.ts]: Sure. But it's good name for the incremental provider method too.

Comment (taras) [reply plugins/catalog-backend-module-incremental-ingestion/src/types.ts]: Yeah, it's a good method name because it makes the purpose very clean and it's recognizable. I find `deltaMapper` kinda confusing.

Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/types.ts]: Done!

File: plugins/catalog-backend-module-incremental-ingestion/src/types.ts (hunk 2 above)
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/types.ts]: Done!

File: plugins/catalog-backend-module-incremental-ingestion/api-report.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/api-report.md]: I think this is fully addressed by the engine, actually. It accepts a payload of both emitted entities and of deleted entities.

**Review by taras — APPROVED** [review]
Looks great

File: plugins/catalog-backend-module-incremental-ingestion/src/types.ts [hunk 5]
```diff
@@ -75,6 +75,22 @@ export interface IncrementalEntityProvider<TCursor, TContext> {
...
   around(burst: (context: TContext) => Promise<void>): Promise<void>;
+
+  /**
+   * If present, this method accepts an incoming event, and maps the
+   * payload to an object containing a delta mutation.
+   *
+   * If a delta is present, the incremental entity provider will apply
+   * it automatically.
+   */
+  onEvent?: (payload: unknown) => {
```
Comment (freben) [inline plugins/catalog-backend-module-incremental-ingestion/src/types.ts:86]: Again, I _really_ think that this should be `params: EventParams` instead of the payload. The other fields in the event params are there for good reason :) You'll quickly run into needing those headers or other things (that are there today or get added in the future).

File: plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts [hunk 6]
```diff
@@ -329,4 +334,69 @@ export class IncrementalIngestionEngine implements IterationEngine {
...
+  async onEvent(params: EventParams): Promise<void> {
+    const { topic, eventPayload } = params;
+    if (topic !== this.providerEventTopic) {
+      return;
+    }
+
+    const { logger, provider, connection } = this.options;
+    const providerName = provider.getProviderName();
+    logger.info(
+      `incremental-engine: Received ${this.providerEventTopic} event`,
```
Comment (freben) [inline plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts:347]: if you had the full EventParams as suggested, you'd get the topic from there instead

Comment (freben) [reply plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts:347]: Also should this be on `info` level? Might be sensible to lower to debug, otherwise it'll get pretty noisy on busy topics

File: plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts [hunk 7]
```diff
@@ -329,4 +334,69 @@ export class IncrementalIngestionEngine implements IterationEngine {
...
+
+      if (update.delta.removed.length > 0) {
+        await this.manager.deleteEntityRecordsByRef(update.delta.removed);
+      }
+
+      await connection.applyMutation({
+        type: 'delta',
+        ...update.delta,
+      });
+      logger.info(
```
Comment (freben) [inline plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts:389]: debug?

File: plugins/catalog-backend-module-incremental-ingestion/src/module/WrapperProviders.ts [hunk 8]
```diff
@@ -75,14 +75,14 @@ export class WrapperProviders {
...
       new IncrementalIngestionDatabaseManager({ client: this.options.client }),
       loggerToWinstonLogger(this.options.logger),
-    );
+    ).createRouter();
   }
 
-  private async startProvider(
-    provider: IncrementalEntityProvider<unknown, unknown>,
+  private async startProvider<TCursor, TContext>(
+    provider: IncrementalEntityProvider<TCursor, TContext>,
```
Comment (freben) [inline plugins/catalog-backend-module-incremental-ingestion/src/module/WrapperProviders.ts:85]: could revert this change now as well, right? Same "T vs unknown" situation as the other one

File: plugins/catalog-backend-module-incremental-ingestion/src/router/routes.ts [hunk 9]
```diff
@@ -15,218 +15,269 @@
...
-        logger.debug(`${provider} - Ingestion record found`);
-        res.json({
-          success: true,
-          message: 'Provider is currently restarting, please wait.',
-        });
-      } else {
-        res.status(404).json({
+    });
+
+    router.post(`${PROVIDER_BASE_PATH}/delta`, async (req, res) => {
```
Comment (freben) [inline plugins/catalog-backend-module-incremental-ingestion/src/router/routes.ts:242]: Hm, is this for debugging? How is this one used?

File: plugins/catalog-backend-module-incremental-ingestion/src/types.ts (hunk 5 above)
Comment (pjungermann) [reply plugins/catalog-backend-module-incremental-ingestion/src/types.ts:86]: I agree. It took me some iterations to end up at EventParams and originally, I only had the payload, too.
I think it is better to support EventParams and stay more aligned.

File: plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts [hunk 10]
```diff
@@ -37,6 +41,7 @@ export class IncrementalIngestionEngine implements IterationEngine {
       { minutes: 30 },
       { hours: 3 },
     ];
+    this.providerEventTopic = `${options.provider.getProviderName()}-push`;
```
[the remark this answers is not in the document]
Comment (pjungermann) [reply plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts:44]: hmm.. I wonder a bit about this topic, too.

Could you describe a bit the flow?

E.g., let's assume I've received a webhook event via HTTP ingress from Bitbucket.org for the event type `repo:push` on the topic `bitbucketCloud` which then gets routed to the topic `bitbucketCloud.repo:push`.

Or from GitHub to topic `github` and then routed to `github.push`.

What would happen next? How would it get to this subscriber?
Would you use yet another event router to route it to this topic name?

Or couldn't we not just configure the topic name to be used, e.g. as part of `IterationEngineOptions`?

File: plugins/catalog-backend-module-incremental-ingestion/src/types.ts (hunk 5 above)
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/types.ts:86]: Will do. It will require some changes to the engine as well, since the engine currently handles the actual event, but those should be minimal.

File: plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts:44]: There is a route for each incremental entity provider for accepting payloads via HTTP ingress. For example, if you have an incremental provider for BitBucket repositories called `BitBucketEntityProvider`, the route would look like `https://[somehost]/api/incremental/providers/BitBucketEntityProvider/event`. It grabs the body of the request and emits it as a payload with a topic of `BitBucketEntityProvider-push`.

File: plugins/catalog-backend-module-incremental-ingestion/src/router/routes.ts (hunk 9 above)
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/router/routes.ts:242]: I'm going to rename `delta` to `event`, FYI.

This route is the mechanism for picking up events targeting the entity provider. It's designed so that every incremental entity provider is already wired up for receiving events (such as those from subscribed webhooks), and only needs to handle the event passed to it in the `onEvent` method.

File: plugins/catalog-backend-module-incremental-ingestion/src/types.ts (hunk 5 above)
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/types.ts:86]: Done.

File: plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts (hunk 6 above)
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts:347]: Switched to debug. Good call.

File: plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts (hunk 7 above)
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts]: Done. :)

File: plugins/catalog-backend-module-incremental-ingestion/src/router/routes.ts (hunk 9 above)
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/router/routes.ts]: This route is gone, now that I'm relying on the events backend.

File: plugins/catalog-backend-module-incremental-ingestion/src/module/WrapperProviders.ts (hunk 8 above)
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/module/WrapperProviders.ts]: Done!

File: plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (dekoding) [reply plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts]: Actually, this route is gone now. Instead, the plugin will be reliant on the presence of the events backend.

**Review by freben — APPROVED** [review]
Let's go :)
